cmd/nknc/commands/asset.go

```diff
@@ -0,0 +1,195 @@
+package commands
+
+import (
+	"encoding/hex"
+	"fmt"
+	"os"
+
+	api "github.com/nknorg/nkn/v2/api/common"
+	"github.com/nknorg/nkn/v2/api/httpjson/client"
+	"github.com/nknorg/nkn/v2/common"
+	"github.com/nknorg/nkn/v2/config"
+	"github.com/nknorg/nkn/v2/util/log"
+	"github.com/nknorg/nkn/v2/vault"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+)
+
+const (
+	RANDBYTELEN = 4
+)
+
+// assetCmd represents the asset command
+var assetCmd = &cobra.Command{
+	Use:   "asset",
+	Short: "asset registration, issuance and transfer",
+	Long:  "",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		showusage := true
+		cmd.Flags().Visit(func(name *pflag.Flag) {
+			showusage = false
+		})
+		if showusage {
+			return cmd.Usage()
+		}
+		if err := assetAction(); err != nil {
+			log.Error(err)
+		}
+		return nil
+	},
+}
+
+var (
+	issue     bool
+	transfer  bool
+	to        string
+	value     string
+	assetname string
+	symbol    string
+	precision uint32
+)
+
+func init() {
+	rootCmd.AddCommand(assetCmd)
+
+	assetCmd.Flags().BoolVarP(&issue, "issue", "i", false, "issue asset")
+	assetCmd.Flags().BoolVarP(&transfer, "transfer", "t", false, "transfer asset")
+	assetCmd.Flags().StringVarP(&walletFile, "wallet", "w", config.Parameters.WalletFile, "wallet name")
+	assetCmd.Flags().StringVar(&walletPassword, "password", "", "wallet password")
+	assetCmd.Flags().StringVar(&to, "to", "", "asset to whom")
+	assetCmd.Flags().StringVarP(&value, "value", "v", "", "asset amount in transfer asset or totalSupply in inssue assset")
+	assetCmd.Flags().StringVarP(&Fee, "fee", "f", "", "transaction fee")
+	assetCmd.Flags().Uint64Var(&Nonce, "nonce", 0, "nonce")
+	assetCmd.Flags().StringVar(&assetname, "name", "", "asset name")
+	assetCmd.Flags().StringVar(&symbol, "symbol", "", "asset symbol")
+	assetCmd.Flags().Uint32Var(&precision, "precision", 0, "asset precision")
+
+	assetCmd.Flags().MarkHidden("password")
+}
+
+func parseAddress() common.Uint160 {
+	if address := to; address != "" {
+		pg, err := common.ToScriptHash(address)
+		if err != nil {
+			fmt.Println("invalid receiver address")
+			os.Exit(1)
+		}
+		return pg
+	}
+	fmt.Println("missing flag [--to]")
+	os.Exit(1)
+	return common.EmptyUint160
+}
+
+func assetAction() error {
+	var err error
+	var txnFee common.Fixed64
+
+	if Fee == "" {
+		txnFee = common.Fixed64(0)
+	} else {
+		txnFee, err = common.StringToFixed64(Fee)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+			return err
+		}
+	}
+
+	var resp []byte
+	switch {
+	case issue:
+		if walletFile == "" {
+			fmt.Println("wallet file is required with [--wallet]")
+			return nil
+		}
+
+		myWallet, err := vault.OpenWallet(walletFile, GetPassword(walletPassword))
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+			os.Exit(1)
+		}
+
+		if assetname == "" {
+			fmt.Println("asset name is required with [--name]")
+			return nil
+		}
+
+		if symbol == "" {
+			fmt.Println("asset symbol is required with [--symbol]")
+			return nil
```

Comment:
L119 returns nil while L125 returns err. Is this expected?

Reply:
this is from the original code. I just took it as it is:
```
		if symbol == "" {
			fmt.Println("asset symbol is required with [--symbol]")
			return nil
		}

		totalSupply, err := common.StringToFixed64(value)
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			return err
		}
```

L125 has an actual error. L119 is just a check if flag is set. this makes sense to me.

Reply:
But the behavior of cobra seems to be different from the library used before. Is it going to show the help text if you return error here? We only want error to be printed, but not help text